### PR TITLE
Return a `Point` object from the PG Point type

### LINKED ISSUE #20441: Columns with arrays of points are not working
It is not possible to have attributes of arrays of points, i.e. columns like `t.column :polygon, :point, array: true`. Assigning them leads to wrongly cast SQL values.

Instead of `{ (x,y), (x,y) }`,   `{ {x,y}, {x,y} }` is sent to the database (which is rejected by PostgreSQL)

This is because the array column type can process nested/multidimensional arrays, which is perfect for most sub types. But for arrays of points the lowest array level belongs to the points and should be handled by the point type and not the array type.

Executable test case: https://gist.github.com/daniel-rikowski/8579c2c591ec01814036

 
I've hacked together two monkey patches for Rails 4 and 5 which modify the array casting methods: For arrays of points the casting stops at the second lowest level and leaves the rest to the sub type. 

I'd put together a pull request, but I'm not sure this is the best solution. It feels wrong to have special handling in the array type for a specific sub type...

### PR DESCRIPTION
**Pull request by sgrif**
This introduces a deprecation cycle to change the behavior of the
default point type in the PostgreSQL adapter. The old behavior will
continue to be available for the immediate future as `:legacy_point`.

The current behavior of returning an `Array` causes several problems,
the most significant of which is that we cannot differentiate between an
array of points, and a point itself in the case of a column with the
`point[]` type.

The attributes API gives us a reasonable way to have a proper
deprecation cycle for this change, so let's take advantage of it. If we
like this change, we can also add proper support for the other geometric
types (line, lseg, box, path, polygon, and circle), all of which are
just aliases for string today.

Fixes #20441

### CODE REVIEW

**Comment by rafaelfranca** [comment]
:shipit:

**Comment by meinac** [comment]
Ups, I was looking at the pr and it has been merged before my comment :smile:

File: activerecord/lib/active_record/connection_adapters/postgresql/oid/rails_5_1_point.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        class Rails51Point < Type::Value # :nodoc:
+          include Type::Helpers::Mutable
+
+          def type
+            :point
+          end
+
+          def cast(value)
+            case value
+            when ::String
```
Comment (mroach) [inline activerecord/lib/active_record/connection_adapters/postgresql/oid/rails_5_1_point.rb]: When an empty string is passed-in, this fails with "can't convert nil into Float". This ends-up happening when a user doesn't enter a value on a form for example. Should this method be converting empty string to nil and storing it as such?

**Pull request merged**